penguin/llm/runtime.py

```
@@ -287,40 +339,95 @@ async def execute_pending_tool_call(
-        import json as _json
+        handler = getattr(api_client, "client_handler", None)
+        plural_getter = getattr(handler, "get_and_clear_pending_tool_calls", None)
+        tool_infos = await _call_pending_tool_getter(plural_getter)
+        if isinstance(tool_infos, list):
+            return [item for item in tool_infos if isinstance(item, dict)]
+        if isinstance(tool_infos, dict):
+            return [tool_infos]
+
+        singular_getter = getattr(handler, "get_and_clear_last_tool_call", None)
+        tool_info = await _call_pending_tool_getter(singular_getter)
+    except Exception:
```

Ruff (0.15.12)

[warning] Do not catch blind exception: `Exception`

(BLE001)